Proposal: add no_std support, with a default 'std' feature

oconnor663 opened the pull request:
I was going to file an issue to discuss `no_std` support, but I figured the changes might be small enough that I could just whip together an example implementation, to make it easier to discuss :) Please let me know if you'd be interested in including something like this. I'd be happy to make changes in this PR, or to turn the whole effort over to you.

oconnor663 committed:
add no_std support, with a default 'std' feature

cesarb commented on src/lib.rs:
```diff
@@ -39,6 +48,7 @@ pub mod blake2b;
 pub mod blake2s;
 
 /// Runs the self-test for both BLAKE2b and BLAKE2s.
+#[cfg(test)]
```
Hiding the selftest functions would change the crate API in the `cfg(std)` case.

oconnor663 replied:
Ah silly me, I didn't pay attention to the fact that they were `pub`. Though if we want to keep them in the API, having them return `ArrayVec` would be a breaking change also. What's the right move?

oconnor663 replied:
Maybe changing the API in the `no_std` case is the right thing to do? All current callers should default to enabling `std`, and so there shouldn't be any breaking changes?

cesarb replied:
The three selftest functions already return nothing, see for instance https://docs.rs/blake2-rfc/0.2.17/blake2_rfc/fn.selftest.html. No need to change the API.

The selftest_seq function is internal, it might be `pub` but is within a non-`pub` module, so its API is not a concern.

cesarb commented on src/blake2.rs:
```diff
@@ -298,13 +301,14 @@ macro_rules! blake2_impl {
 #[cold]
-pub fn selftest_seq(len: usize) -> Vec<u8> {
-    use std::num::Wrapping;
+pub fn selftest_seq(len: usize) -> ::arrayvec::ArrayVec<[u8; 65536]> {
```
The maximum number of elements used by the selftest is 1024; it might be better to change `test_write()` to also use 1024 bytes, instead of always allocating 64k on the stack here.

oconnor663 replied:
Sure, should we make `test_write` just loop a thousand times or something like that?

cesarb replied:
I don't think it's needed, since `test_write` is there just to check that `write` does the same as `update`. As long as the input is bigger than the hash block size, it should be enough. In that particular testcase, `selftest_seq` is just a source of pseudorandom data, what it generates doesn't matter as long as it's not a stream of zeros.

cesarb reviewed (changes requested):
Looks good, thanks. The only issue is that since the three selftest functions are exported as part of the crate's public interface, removing them would needlessly break compatibility. I see no problem having arrayvec as a plain dependency, instead of a dev-dependency; if that's an issue, we could add another default feature for the selftests.

oconnor663 committed:
ignore the 4GB tests in debug mode

oconnor663 commented:
I squashed in the changes you suggested. I also added a second commit to `ignore` the 4GB tests in debug mode. On my machine at least they took long enough that running the debug tests wasn't feasible. This way at least the user gets a nice printout making it clear they're being skipped. Let me know what you think of that one.

cesarb commented:
Nice trick with the 4GB tests! Having to always use --release to run `cargo test` was an annoyance.

oconnor663 commented:
Thanks!
